ext/bigdecimal/bigdecimal.c

```diff
@@ -7192,6 +7193,46 @@ VpItoV(Real *m, SIGNED_VALUE ival)
 }
 #endif
 
+/*
+ * Calculate sqrt for the required digits
+ */
+static void
+VpSqrtNewton(Real *f, Real *r, Real *x, Real *y, size_t digits)
+{
+    size_t y_maxprec;
+    if (digits < 15) {
+        double val;
+        SIGNED_VALUE e, n;
+        VpVtoD(&val, &e, x);  /* val <- x  */
+        e /= (SIGNED_VALUE)BASE_FIG;
+        n = e / 2;
+        if (e - n * 2 != 0) {
+            val /= BASE;
+            n = (e + 1) / 2;
+        }
+        VpDtoV(y, sqrt(val)); /* y <- sqrt(val) */
+        y->exponent += n;
+        return;
+    }
+
+    // y0 = (1 ± ε) * √x
+    // y = (y0 + x/y0) / 2 ≒ (1 ± ε**2 / 2) * √x
+    // Requirements: ε**2 / 2 < 10**(-digits)
+    // ε = 10**(-(digits+1)/2) is enough
+    // Required digits for y0 should be at least (digits+1)/2+1
+    // Adding some safe margin because it does not affect performance
+    VpSqrtNewton(f, r, x, y, 3 + digits / 2);
```

Comment:
Changing this to `(digits+1) / 2 + 1` still passes this test.
```ruby
10000.times do
  x=BigDecimal("#{1+rand}e#{rand(-100..100)}")
  prec=rand(10..1000)
  raise unless (x.sqrt(prec)/x.sqrt(prec+100)-1).exponent <= -prec-10
end
```
(Initially, it failed in some case because there was a bug in VpVtoD)